torch_geometric/explain/algorithm/gnn_explainer.py

```diff
@@ -0,0 +1,539 @@
+        **kwargs,
+    ):
+        super().__init__()
+        self.epochs = epochs
+        self.lr = lr
+        self.coeffs.update(kwargs)
+        # node_mask will never actually be None
+        self.node_mask = None
+        # edge_mask is only None if edge_mask_type is != MaskType.object
+        self.edge_mask = None
+
+    def forward(
+        self,
+        model: torch.nn.Module,
+        x: Tensor,
+        edge_index: Tensor,
+        explainer_config: ExplainerConfig,
+        model_config: ModelConfig,
+        target: Tensor,
+        target_index: Optional[Union[int, Tensor]] = None,
+        node_index: Optional[int] = None,
+        **kwargs,
+    ) -> Explanation:
+
+        if isinstance(target_index, Tensor) and target_index.numel() > 1:
+            raise ValueError(
+                "GNNExplainer only supports single target index for now")
+
+        assert model_config.task_level in [
+            ModelTaskLevel.graph, ModelTaskLevel.node
+        ]
+
+        model.eval()
+
+        if model_config.task_level == ModelTaskLevel.node:
+            node_mask, edge_mask = self._explain_node(model, x, edge_index,
+                                                      explainer_config,
+                                                      model_config, target,
+                                                      node_index, target_index,
+                                                      **kwargs)
+        elif model_config.task_level == ModelTaskLevel.graph:
+            node_mask, edge_mask = self._explain_graph(model, x, edge_index,
+                                                       explainer_config,
+                                                       model_config, target,
+                                                       target_index, **kwargs)
+
+        if explainer_config.node_mask_type == MaskType.object:
+            node_feat_mask = None
+        else:
+            node_feat_mask = node_mask
+            node_mask = None
+
+        self._clean_model(model)
+
+        # build explanation
+        return Explanation(x=x, edge_index=edge_index, edge_mask=edge_mask,
+                           node_mask=node_mask, node_feat_mask=node_feat_mask)
+
+    def _explain_graph(
+        self,
+        model: torch.nn.Module,
+        x: Tensor,
+        edge_index: Tensor,
+        explainer_config: ExplainerConfig,
+        model_config: ModelConfig,
+        target: Tensor,
+        target_index: Optional[Union[int, Tensor]] = None,
+        **kwargs,
+    ) -> Tuple[Tensor, Optional[Tensor]]:
+        self._train_node_edge_mask(model, x, edge_index, explainer_config,
+                                   model_config, target, target_index, None,
+                                   **kwargs)
+
+        node_mask = self.node_mask.detach().sigmoid().squeeze(-1)
+        if explainer_config.edge_mask_type == MaskType.object:
+            edge_mask = self.edge_mask.detach().sigmoid()
+        else:
```

Comment:
In a couple other areas, you do
```python
if mask_type == A:
    ...
elif mask_type == B:
    ...
```

The benefit of that is seeing clearly what triggers the second clause, but the downside is an invalid input might cause no action. Would if/elif/else raise Exception() be a better framework? Regardless I think it would be good to standardize.

Reply:
yes I could uniformize that, it's a bit shaky at the moment (this also relates to your previous comment on the assert being handled in the `supports` function)

Reply:
Sorry, could you point me to the precise loop you're referring ? I can't seem to find it.